### PR TITLE
Add bignum test case generation script

### LINKED ISSUE #5921: Create a script for generating bignum tests
Create a script for generating unit tests in a systematic way. This is similar to `tests/scripts/generate_psa_tests.py`.

To start with, generate a few bignum simple test cases (into a new file `test_suite_bignum.generated.data`) for one or more existing test function, just to get the framework started. There will be follow-ups to generate useful bignum and crypto test cases.

Make sure that the new test data file is taken into account wherever they need to be. New data files will be added later, so call `tests/scripts/generate_bignum_tests.py --list`, don't hard-code the list of generated files if not absolutely necessary. Check where `generate_psa_tests.py` is called; `generate_bignum_tests.py` will likely need to be called in the same places. In particular, both with make (`**/Makefile`) and with CMake (`**/CMakeLists.txt`):

* Building the unit test must build the new test suite.
* Running the unit tests must run the new test suite.
* `make generated_files` must generate the new data file.
* `tests/scripts/check_generated_files.sh` must take the new generated file into account.
* In 2.28, the new file must be checked into git. In development, it must be `.gitignore`d.

### PR DESCRIPTION
**Pull request by wernerlewis**
## Description
Resolves #5921

Adds a test case generation framework for bignum tests, similar to `generate_psa_tests.py`. This PR adds a few bignum arithmetic test cases to show the usage of the test framework, and it will be expanded to cover more bignum/ECC functionality in subsequent PRs.

This PR should ideally be merged after #6070. The temporary radix argument commit can then be removed before merging this PR.

## Status
**READY**

## Requires Backporting

Yes 2.28

### CODE REVIEW

**Review by minosgalanakis — DISMISSED** [review]
Overall looks good, and will do the job it is supposed to do. My main concerns are:

- Complexity. The code may be hard to maintain, especially since we are combining iterator and class generation. It assumes the reviewer is quite apt with modern Python.
- Error handling. A lot of edge cases rely on assumed correct input, especially in earlier stages of inheritance.

Would having a document or a comment block descibing how you can easily add new tests be benefical?

File: tests/scripts/generate_bignum_tests.py
Comment (gilles-peskine-arm) [inline tests/scripts/generate_bignum_tests.py:156]: Isn't this, and `input_cases`, an iterator over pairs of strings? So `Iterator[Tuple[str, str]]`.

File: tests/scripts/generate_bignum_tests.py
Comment (gilles-peskine-arm) [inline tests/scripts/generate_bignum_tests.py:161]: My natural inclination would be to write
```
yield from itertools.combinations(cls.input_vals, 2)
yield from cls.input_cases
```
What's the difference?

File: tests/scripts/generate_bignum_tests.py
Comment (gilles-peskine-arm) [inline tests/scripts/generate_bignum_tests.py:197]:
```suggestion
    """Target for absolute value comparison."""
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

File: tests/scripts/generate_bignum_tests.py
Comment (yanesca) [reply tests/scripts/generate_bignum_tests.py:70]: I think that having numbered test is useful and is worth the cost of that 3-4 characters in the title. Comparing test results across different commits is only a problem if we added test cases or changed generation order between the two commits. Also, in the overwhelming majority of cases we are looking at the tests within a single commit, where having a number makes identifying the failing test cases much easier.

Could we please keep the test case numbering?

Comment (gilles-peskine-arm) [reply tests/scripts/generate_bignum_tests.py:70]: There is value in having an easy-to-compare unique identifier on test cases. That unique identifier doesn't need to be meaningful, so it can be a number. Now the question is, what kind of stability do we expect on this number?

At one extreme, we can just use the line number. This changes pretty much whenever we add or remove test cases (or comments). Nice properties: lookup by number is extremely easy; the numbers are increasing (but not consecutive); the numbering can be done fully automatically by the test framework.

At another extreme, we can assign a number (or other short identifier) when we write a test case, and change to it. Advantage: the unique identifier is permanent. Downside: has to be managed manually, may cause conflicts if two concurrent tracks of work add the same unique identifier to different test cases.

With the intermediate solution we'd be introducing here, test case numbers would apply only to automatically generated test cases, so this isn't a general solution. We can easily tweak the scope of the numbering (per test function, per data file…) but any scope has its own downside (per test function: makes it hard to figure out what the unique part is; per test suite: makes the numbers unstable when we add unrelated test cases). I think this is not a good design for test case numbering. So I maintain my opposition.

I think the idea of test numbering is a good thing. I, too, have spent time copy-pasting descriptions around to search them, and sometimes looking at the wrong test case because I missed a difference in the description. So far it's been low on my tech debt list, but seeing I'm not the only one interested, then by all means let's add it to our tech debt. But [let's first agree on a design](https://github.com/Mbed-TLS/mbedtls/issues/6227), and introduce it as its own feature rather than part of some unrelated work.

Comment (yanesca) [reply tests/scripts/generate_bignum_tests.py:70]: I am happy to have discussion about the design, but could we please not block this PR with that discussion?

This is a test framework, whatever we decide, won't affect the test cases we add while discussing the matter. I really would like to start using this in the Bignum refactoring work as soon as possible. We could tackle the global numbering problem in a different PR.

Comment (gilles-peskine-arm) [reply tests/scripts/generate_bignum_tests.py:70]: > could we please not block this PR with that discussion?

Agreed, absolutely. Test case numbers are going to cause merge conflicts if we work on this script in parallel, which is likely in the near future. So let's not put them in now, and add them later if we decide on that strategy.

Comment (yanesca) [reply tests/scripts/generate_bignum_tests.py:70]: Great, thanks, sounds good!

Regarding the temporary presence of test numbering in automated tests:

The script output is generated files, they won't be committed and won't cause merge conflicts. If we work in parallel on the script, we won't change the part that adds the numbering and the other PR will likely remove it. Again, it seems as a clean merge, I can't see the conflict.

On the Bignum refactoring side, while we don't have the global numbering solution, having this local numbering would be really helpful. Some tests have hundreds of test cases, it is very difficult to navigate them without any numbering.

Comment (gilles-peskine-arm) [reply tests/scripts/generate_bignum_tests.py:70]: > The script output is generated files, they won't be committed and won't cause merge conflicts

Well, I was thinking of 2.2x, where they are committed. But it's true that on 2.2x, we typically have quick cycle from making the backport to merging it, so the risk of conflicts is small.

Comment (tom-cosgrove-arm) [reply tests/scripts/generate_bignum_tests.py:70]: Is this going to be backported? I know it says "Yes" in the description, but this is clearly a new feature. And many of the newly-generated tests won't be applicable to bignum-2.28. Of course, having them the same has advantages, but I thought we'd agreed that bignum-development and bignum-2.28 would diverge?

Comment (yanesca) [reply tests/scripts/generate_bignum_tests.py:70]: This is a test framework, which is helpful to have on both and the test framework doesn't have to (or shouldn't) diverge. Also, this would make backporting tests easier. (We are adding more tests to other areas as well, not just the new code.)

File: tests/scripts/generate_bignum_tests.py
Comment (wernerlewis) [reply tests/scripts/generate_bignum_tests.py]: I've replaced this with a `__new__` override as it is tidier, and this avoids the need for `super().__init__()` in all derived classes, which could be an issue with deeper inheritance.

File: tests/scripts/generate_bignum_tests.py
Comment (wernerlewis) [reply tests/scripts/generate_bignum_tests.py]: This could be set in `BaseTarget` for now, but for example if `ECP` cases are added in future, this would require rework to support tests being written to a different target file. With this approach `ECPTarget` can be added, and used to derive ECP test cases from. This approach allows reuse of the existing `TestGenerator` class fairly easily, although there are benefits to other approaches.

File: tests/scripts/generate_bignum_tests.py
Comment (wernerlewis) [reply tests/scripts/generate_bignum_tests.py]: I've updated the naming of these attributes and added details in the docstrings to clarify these.

File: tests/scripts/generate_bignum_tests.py
Comment (wernerlewis) [reply tests/scripts/generate_bignum_tests.py]: This is unnecessary, I've removed `@property` in these cases.

File: tests/scripts/generate_bignum_tests.py
Comment (wernerlewis) [reply tests/scripts/generate_bignum_tests.py]: Yes this is similar to how `unittest` works. I do think that this can become complicated, especially with the mixing of both `TestCase` generation for the class, and the enumeration of derived classes. I've split this functionality so that `generate_tests()` will just handle the enumeration of derived classes, and call `generate_function_tests()` if `test_function` is set. That will then handle the `TestCase` generation. I think this reduces complexity when adding new classes, as only the `TestCase` generation needs to be implemented, rather than needing to handle the enumeration as well.

These changes should also support intermediate classes with common elements, as no test cases will be generated unless `test_function` is set. For adding additional modules, it should be possible to do this by adding a new class such as `ECPTarget` to represent the module, and then deriving from this, although there are other approaches that may be better suited.

File: tests/scripts/generate_bignum_tests.py
Comment (wernerlewis) [reply tests/scripts/generate_bignum_tests.py]: This is correct, in these two cases mypy assumes the type `Iterator[Tuple[str, ...]]` even if the tuple is always of a fixed-width. I've added casts for these so that the correct type can be used.

File: tests/scripts/generate_bignum_tests.py
Comment (wernerlewis) [reply tests/scripts/generate_bignum_tests.py]: There's no difference, I've changed this to the two `yield from` as it is clearer.

File: tests/scripts/generate_bignum_tests.py
Comment (minosgalanakis) [inline tests/scripts/generate_bignum_tests.py:8]: This is not enforced but it could make sense to implement docstings in files after the license using the `__doc__ = """ ...` format.

File: scripts/mbedtls_dev/test_generation.py
Comment (gilles-peskine-arm) [reply scripts/mbedtls_dev/test_generation.py:61]: I don't understand why we'd want anything fancier here.

Also please keep in mind that the average maintainer of this file is primarily a C programmer, and not a Python expert.

File: tests/scripts/generate_bignum_tests.py
Comment (gilles-peskine-arm) [reply tests/scripts/generate_bignum_tests.py:8]: Why?

In any case the file description (docstring in Python) goes before the license (like in [`*.h`](https://mbed-tls.readthedocs.io/en/latest/kb/development/mbedtls-coding-standards/#header-files) and [`*.c`](https://mbed-tls.readthedocs.io/en/latest/kb/development/mbedtls-coding-standards/#source-files)).

File: tests/scripts/generate_bignum_tests.py
Comment (gilles-peskine-arm) [reply tests/scripts/generate_bignum_tests.py:181]: I'm not seeing the deviation from [PEP8](https://peps.python.org/pep-0008/#documentation-strings) here: a single-line docstring has the closing quotes on the same line, a multi-line docstring has the closing quotes on a separate line. (We do deviate from the second rule in other places though, but not in this PR.)

File: tests/scripts/generate_bignum_tests.py
Comment (gilles-peskine-arm) [reply tests/scripts/generate_bignum_tests.py:211]: There's a strong unifying theme here, so I don't see any value in splitting. We may want to reorganize some things when we start generating ECC arithmetic tests, by creating a common module used by both bignum and ecc test generators, but I think it's too early to tell what code we'll want to share.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
A couple of minor documentation issues, and an unnecessary cast. Other than that looks good to me.

File: scripts/mbedtls_dev/test_generation.py
Comment (wernerlewis) [reply scripts/mbedtls_dev/test_generation.py]: I've rewritten this comment to describe what is being added to `self.targets` more clearly, and removed the other uses of "Targets".

File: tests/scripts/generate_bignum_tests.py
Comment (minosgalanakis) [reply tests/scripts/generate_bignum_tests.py:190]: I meant about the whitespace after the closing of the string.

~~~~~
def some_function(void):
""" That is a short description ""

     x =5
~~~~~

vs 

~~~~~
def some_function(void):
""" That is a short description ""
     x =5
~~~~~

Bot are acceptable but it is better to be consistent

File: scripts/mbedtls_dev/test_generation.py
Comment (minosgalanakis) [reply scripts/mbedtls_dev/test_generation.py:64]: This was more of a proposal as a stepping stone if we are aiming to intergrate those two into the future. Not a strong ask by any means ;)

File: tests/scripts/generate_bignum_tests.py
Comment (minosgalanakis) [reply tests/scripts/generate_bignum_tests.py:8]: The idea behind docstrings was to have accessible manuals in the even during the interpreter execution runtime. e.g `print(__doc__)` . 

Because those file attributes are set on the beggining of the file, it is common to use intermediate variables when including large licence or docstrings

e.g 

~~~~~
"""random_file.py: Some dosctring describing it"""

__license__ = "GPLv2"
__version__ = "1.0"
__author__  = "A Developer"
__email__   = "developer@mail.com"
~~~~~

Or

~~~~~
__doc__     = _dc_
__license__ = "GPLv2"
__version__ = "1.0"
__author__  = "A Developer"
__email__   = "developer@mail.com"

_dc_ = """ random_file.py: Some multiline
                 dosctring describing it 
           """
~~~~~

Anyhow we do not use that format, and it is extremely unlikely to scale our scripts to the size that it would be required. 

Could it be worth to mention in our guidelines ?

File: tests/scripts/generate_bignum_tests.py
Comment (minosgalanakis) [inline tests/scripts/generate_bignum_tests.py:238]: This is precicely why using __doc__ was established. The same can be achieved by 

~~~~~
__short_doc __ = """Generate test data for bignum functions. ""
__file_help__ = """With no arguments, generate all test data. With non-option arguments,
generate only the specified files.

Class structure:

Child classes of test_generation.BaseTarget (file targets) represent an output
"""

.....
.....
.....

test_generation.main(sys.argv[1:], __file_help__ ))

~~~~~

This shall not block this PR, this is just for the purposes of discussion

**Review by minosgalanakis — APPROVED** [review]
Looks good to me.

File: tests/scripts/generate_bignum_tests.py
Comment (gilles-peskine-arm) [inline tests/scripts/generate_bignum_tests.py:80]: Non-blocker: this information is relevant when reading the function's code, but not when using the function. So it should be a comment, not part of the documentation string.

File: scripts/mbedtls_dev/test_generation.py
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/test_generation.py:194]: Well, not quite. `--directory` is still interpreted relative to the root of the mbedtls tree. An `abspath` call is missing somewhere (or alternatively we could stop using `chdir`, but that would be annoying).

But since this is preexisting (I know it worked when I originally wrote it, but I might have broken it before I even committed), this is a non-blocker.

**Review by gilles-peskine-arm — APPROVED** [review]
Approved. I have two comments but neither are blockers.

In the interest of making it possible to use this script in the ongoing bignum work as soon as possible, I intend to merge this as soon as CI passes. We still need to make a 2.28 backport (which should be an identical script, plus in 2.28 we commit the output into version control).

File: scripts/mbedtls_dev/test_generation.py
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/test_generation.py:1]: I only just realized, but we already have test _code_ generation (`tests/scripts/generate_test_code.py`, turns `.function` files into `.c`), so “test generation” is ambiguous: we should specify that this is test _data_ generation. I propose (in a follow-up) to rename this module to `test_data_generation`.

File: scripts/mbedtls_dev/test_generation.py
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/test_generation.py:30]: These should be relative imports. No reason to assume that `scripts` is in the search path.

Again, this is a preexisting bug which doesn't matter in our normal usage, so not a blocker.

File: tests/scripts/generate_bignum_tests.py
Comment (gilles-peskine-arm) [reply tests/scripts/generate_bignum_tests.py:80]: https://github.com/Mbed-TLS/mbedtls/pull/6296 has fixes for this and the other non-blocker issues I raised in this review.

**Pull request merged**
